File: 354-RussianDollEnvelopes/solution.cc
/**
 * 354. Russian Doll Envelopes
 * 
 * Chang Li @ Mountain view
 * Jul. 13, 2016
 */
class Solution {
public:

    int maxEnvelopes(vector<pair<int, int>>& envelopes) {
        // Sort (Cumtomerized Comp func is necessary)
        sort(envelopes.begin(), envelopes.end(),
                [](const pair<int, int>& a, const pair<int, int>& b) {
                    if (a.first == b.first) return a.second > b.second;
                    else return a.first < b.first;
                });
        // Find Longest Increasing Sequence
        vector<int> table;
        for (auto p : envelopes) {
            auto it = lower_bound(table.begin(), table.end(), p.second);
            if (it == table.end()) table.push_back(p.second);
            else *it = p.second;
        }
        return table.size();
    }
};

File: 354-RussianDollEnvelopes/solution.py
# Check https://leetcode.com/problems/russian-doll-envelopes/discuss/82761/Python-O(nlogn)-O(n)-solution-beats-97-with-explanation
class Solution:
    def maxEnvelopes(self, envelopes: List[List[int]]) -> int:
        envelopes.sort(key=lambda e: (e[0], -1 * e[1]))
        best_h = []
        for w, h in envelopes:
            pos = bisect.bisect_left(best_h, h)
            if pos == len(best_h):
                best_h.append(h)
            else:
                best_h[pos] = h
        return len(best_h)
